README.md
# Bolt Common

This library provides utility functions to help simplify menial tasks.

The Bolt team believes the PHP error reporting system is a mistake. Many
built-in functions utilize it, leading to inconsistent results and head
scratching.

This library provides some wrappers around some of these functions. Our code
should always throw exceptions instead of triggering errors/warnings/notices
(excluding deprecation warnings).

Table of Contents:
- [Assert](#assert)
- [Deprecated](#deprecated)
- [Ini](#ini)
- [Json](#json)
- [Serialization](#serialization)
- [Str](#str)

-----

## `Assert`

Additional assertions built on `Webmozart\Assert`


### `isArrayAccessible`

Throws `InvalidArgumentException` if `$value` is not an array or object
implementing `ArrayAccess`.

```php
isArrayAccessible($value, string $message = ''): void
```


### `isInstanceOfAny`

Throws `InvalidArgumentException` if `$value` is not an instance of one of the
given classes/interfaces.

```php
isInstanceOfAny($value, string[] $classes, string $message = ''): void
```

### `isIterable`

Throws `InvalidArgumentException` if `$value` is not an _iterable_. Same as
`isTraversable()`, just a better name.

```php
isIterable($value, string $message = ''): void
```


## `Deprecated`

Helper methods for triggering deprecation warnings.


### `warn`

Shortcut for triggering a deprecation warning for something.

```php
warn(string $subject, string|float $since = null, string $suggest = ''): void
```

Examples:

```php
// Triggers warning: "Doing foo is deprecated."
Deprecated::warn('Doing foo');

// Triggers warning: "Doing foo is deprecated since 3.3 and will be removed in 4.0."
Deprecated::warn('Doing foo', 3.3);

// Triggers warning: "Doing foo is deprecated since 3.3 and will be removed in 4.0. Do bar instead."
Deprecated::warn('Doing foo', 3.3, 'Do bar instead');
```


### `method`

Shortcut for triggering a deprecation warning for a method.

```php
method(string|float $since = null, string $suggest = '', string $method = null): void
```

`$suggest` can be a sentence describing what to use instead. Or it can be a
method name or `class::method` which will be converted to a sentence.

`$method` defaults to the method/function it was called from.
  - If called from constructor, warning message says the class is deprecated.
  - If called from magic method, warning message says the method/property
    called with is deprecated.

Example:

```php
class Foo
{
    public function world()
    {
        // Triggers warning: "Foo::world() is deprecated since 3.3 and will be removed in 4.0. Use hello() instead."
        Deprecated::method(3.3, 'hello');
    }
}
```


### `cls`

Shortcut for triggering a deprecation warning for a class.

```php
cls(string $class, string|float $since = null, string $suggest = null): void
```

`$suggest` can be a sentence describing what to use instead. Or it can be a
class name which will be converted to a sentence.

Examples:

```php
// Triggers warning: "Foo\Bar is deprecated."
Deprecated::cls('Foo\Bar');

// Triggers warning: "Foo\Bar is deprecated. Use Bar\Baz instead."
Deprecated::cls('Foo\Bar', null, 'Bar\Baz');
```


## `Ini`

Handles setting and retrieving values from PHP's configuration.


### `has`

Checks whether the key exists.

```php
has(string $key): bool
```


### `getStr`

Get a string value. The default is returned if the key does not exist or the
value is empty.

```php
getStr(string $key, string $default = null): ?string
```


### `getBool`

Get a boolean value. False is returned if the key does not exist or the value
is empty.

```php
getBool(string $key): bool
```


### `getNumeric`

Get a numeric value. The default is returned if the key does not exist or the
value is empty.

```php
getNumeric(string $key, int|float $default = null): int|float|null
```


### `getBytes`

Get a memory size value, such as `memory_limit`, and convert it to an integer.
The default is returned if the key does not exist or the value is empty.

```php
getBytes(string $key, int $default = null): ?int
```


### `set`

Set a new value for the given key.
Throws `RuntimeException` if the key does not exist, it is not editable, or
something goes wrong.

```php
set(string $key, ?scalar $value): void
```


## `Json`

Handles JSON parsing/dumping with error handling.


### `parse`

Parses JSON _string_ to _array_ or _scalar_.
Throws `ParseException` if anything goes wrong.

```php
parse(string $json, int $options = 0, int $depth = 512): string
```

We use [`seld/jsonlint`](https://github.com/Seldaek/jsonlint) to determine why
the parsing failed and include it in the exception message.


### `dump`

Dumps _mixed_ to JSON _string_. Throws `DumpException` if anything goes wrong.

```php
dump(mixed $data, int $options = JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE, int $depth = 512): string
```

If input contains invalid UTF-8 characters we try to convert these for you
before failing.


### `test`

Returns whether the string is valid JSON.

```php
test(string $json): bool
```


## `Serialization`

Handles PHP serialization parsing/dumping with error handling.


### `parse`

Parses PHP serialized _string_.

Throws `ParseException` if a serialized class cannot be found or anything else
goes wrong.

```php
parse(string $value, array $options = []): mixed
```

Note: `$options` parameter is ignored on PHP 5.

See [`unserialize()`](http://php.net/manual/en/function.unserialize.php) for
details.


### `dump`

Dumps anything to a PHP serialized _string_.

Throws `DumpException` if the input is not serializable or anything else goes
wrong.

```php
dump(mixed $value): string
```


## `Str`

Common string methods.


### `replaceFirst`

Replaces the first occurrence of the $search text on the $subject.

```php
replaceFirst(string $subject, string $search, string $replace, bool $caseSensitive = true): string
```


### `replaceLast`

Replaces the last occurrence of the $search text on the $subject.

```php
replaceLast(string $subject, string $search, string $replace, bool $caseSensitive = true): string
```


### `removeFirst`

Removes the first occurrence of the $search text on the $subject.

```php
removeFirst(string $subject, string $search, bool $caseSensitive = true): string
```


### `removeLast`

Removes the last occurrence of the $search text on the $subject.

```php
removeLast(string $subject, string $search, bool $caseSensitive = true): string
```


### `splitFirst`

Splits a $subject on the $delimiter and returns the first part.

```php
splitFirst(string $subject, string $delimiter): string
```


### `splitLast`

Splits a $subject on the $delimiter and returns the last part.

```php
splitLast(string $subject, string $delimiter): string
```

### `endsWith`

Returns whether the subjects ends with the search string.

```php
endsWith(string $subject, string $search, bool $caseSensitive = true): bool
```


### `className`

Returns the class name without the namespace, of a string FQCN, or object.

```php
className(string|object $class): string
```


### `humanize`

Converts a string from camel case and snake case to a human readable string.

```php
humanize(string $text): string
```


### `camelCase`

Converts a string from snake case to camel case.

```php
camelCase(string $text, bool $lowercaseFirstChar = false): string
```


### `snakeCase`

Converts a string from camel case to snake case.

```php
snakeCase(string $text): string
```


CHANGELOG.md
CHANGELOG for Bolt Common
=========================

1.1.0
-----

Released 2017-08-10 ([commits since v1.0.0](https://github.com/bolt/common/compare/v1.0.0...v1.1.0))

 - PR [#17](https://github.com/bolt/common/pull/17)
   - Change: Allow call stack index to be passed into `Deprecated::method` subject parameter
   - Change: Magic call methods now suggest corresponding method on suggested class
   - Change: Removed functionality for magic properties (i.e. __get)
   - Change: Refactored stacktrace logic into separate method
 - PR [#16](https://github.com/bolt/common/pull/16)
   - Added: `Json` constant shortcut flags
   - Change: `Json` to escape line terminators
   - Change: `Json` doesn't pretty print by default
   - Fixed: `Json` will throw an exception for empty strings in PHP 5
 - PR [#15](https://github.com/bolt/common/pull/15)
   - Fixed: `Deprecated::method` use in constructors
 - PR [#14](https://github.com/bolt/common/pull/14)
   - Change: `Deprecated::method` now suggests the method of suggested class

1.0.0
-----

Released 2017-08-01 ([commits](https://github.com/bolt/common/compare/253f473f479d8aa149574fd1ab237da0e9c995c0...v1.0.0))

Change summary:

 - Added: `Assert` class
 - Added: `Json` class
 - Added: `String` class
 - Added: `Deprecated` class
 - Added: `Thrower` class


src/Deprecated.php
<?php

namespace Bolt\Common;

/**
 * This class provides shortcuts for triggering deprecation warnings for various things.
 *
 * @author Carson Full <carsonfull@gmail.com>
 */
class Deprecated
{
    /**
     * Shortcut for triggering a deprecation warning for a method.
     *
     * Example:
     *     class Foo
     *     {
     *         public function hello() {}
     *
     *         public function world()
     *         {
     *             Deprecated::method(3.3, 'hello');
     *         }
     *     }
     * Will trigger: "Foo::world() is deprecated since 3.3 and will be removed in 4.0. Use hello() instead."
     *
     * @param float|null $since   The version it was deprecated in
     * @param string     $suggest A method or class or suggestion of what to use instead.
     *                            If it is a class and the class has a matching method name,
     *                            that will be the suggestion.
     * @param string|int $subject The method or class name or the index of the call stack to reference
     */
    public static function method($since = null, $suggest = '', $subject = 0)
    {
        if ($subject === null || is_int($subject)) {
            list($subject, $function, $class, $constructor) = static::getCaller($subject ?: 0);
        } else {
            Assert::stringNotEmpty($subject, 'Expected a non-empty string. Got: %s');
            $function = $subject;
            $constructor = false;
        }

        // Shortcut for suggested method
        if ($suggest && preg_match('/\s/', $suggest) === 0) {
            // Append () if it is a method/function (not a class)
            if (!class_exists($suggest)) {
                $suggest .= '()';
            } elseif (!$constructor && method_exists($suggest, $function)) {
                // $suggest is class that has matching method name and is not the constructor
                $suggest = $suggest . '::' . $function . '()';
            }
            $suggest = "Use $suggest instead.";
        }

        if (!$constructor) {
            $subject .= '()';
        }

        static::warn($subject, $since, $suggest);
    }

    /**
     * Get info about caller at index.
     *
     * @param int $index
     * @param int $offset
     *
     * @return array [string repr, function name, class name or false, isConstructor]
     */
    protected static function getCaller($index, $offset = 1)
    {
        Assert::greaterThanEq($index, 0);

        $index += $offset + 1;

        $stack = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, $index + 1);
        if (!isset($stack[$index])) {
            throw new \OutOfBoundsException(sprintf('%s is greater than the current call stack', $index - $offset - 1));
        }
        $frame = $stack[$index];

        if (!isset($frame['class'])) {
            // Assert the function isn't called directly from a script,
            // else we would be saying "require() is deprecated" lol.
            if (!function_exists($frame['function'])) {
                $frame = $stack[$index - $offset];
                throw new \InvalidArgumentException(
                    sprintf('%s::%s() must be called from within a function/method.', $frame['class'], $frame['function'])
                );
            }

            return [
                $frame['function'],
                $frame['function'],
                false,
                false,
            ];
        }

        $class = $frame['class'];
        $function = $frame['function'];
        $constructor = $function === '__construct';

        if ($function === '__call' || $function === '__callStatic') {
            $frame = debug_backtrace(false, $index + 1)[$index]; // with args
            $function = $frame['args'][0];
        }

        return [
            $class . (!$constructor ? '::' . $function : ''),
            $function,
            $class,
            $constructor,
        ];
    }

    /**
     * Shortcut for triggering a deprecation warning for a class.
     *
     * @param string     $class   The class that is deprecated
     * @param float|null $since   The version it was deprecated in
     * @param string     $suggest A class or suggestion of what to use instead
     */
    public static function cls($class, $since = null, $suggest = null)
    {
        if ($suggest && preg_match('/\s/', $suggest) === 0) {
            $suggest = ltrim($suggest, '\\');
            $suggest = "Use $suggest instead.";
        }
        $class = ltrim($class, '\\');

        static::warn($class, $since, $suggest);
    }

    /**
     * Shortcut for triggering a deprecation warning for a subject.
     *
     * Examples:
     *     Deprecated::warn('Doing foo');
     *     // triggers warning: "Doing foo is deprecated."
     *
     *     Deprecated::warn('Doing foo', 3.3);
     *     // triggers warning: "Doing foo is deprecated since 3.3 and will be removed in 4.0."
     *
     *     Deprecated::warn('Doing foo', 3.3, 'Do bar instead');
     *     // triggers warning: "Doing foo is deprecated since 3.3 and will be removed in 4.0. Do bar instead."
     *
     * @param string     $subject The thing that is deprecated
     * @param float|null $since   The version it was deprecated in
     * @param string     $suggest A suggestion of what to do instead
     */
    public static function warn($subject, $since = null, $suggest = '')
    {
        $message = $subject . ' is deprecated';

        if ($since !== null) {
            $since = (string) $since;
            $message .= sprintf(' since %.1f and will be removed in %s.0', $since, $since[0] + 1);
        }

        $message .= '.';

        if ($suggest) {
            $message .= ' ' . $suggest;
        }

        static::raw($message);
    }

    /**
     * Trigger a deprecation warning.
     *
     * @param string $message The raw message
     */
    public static function raw($message)
    {
        @trigger_error($message, E_USER_DEPRECATED);
    }
}
